File: src/main/kotlin/khitiara/ffxiv/botflox/xivapi/ClassJobs.kt
package khitiara.ffxiv.botflox.xivapi

object ClassJobs {
    interface ClassJob {
        val id: Int
        val icon: String
        val abbrev: String
    }

    @Suppress("USELESS_CAST")
    fun values(): Collection<ClassJob> = arrayOf(
        CombatClass.values(),
        CombatJob.values(),
        CombatLimitedJob.values(),
        HandClass.values(),
        LandClass.values()
    ).flatten().map { it as ClassJob }

    private val idMap: Map<Int, ClassJob> = values().map { it.id to it }.toMap()
    fun byId(id: Int): ClassJob? = idMap[id]

    enum class CombatClass(override val id: Int, override val icon: String, override val abbrev: String) : ClassJob {
        Gladiator(1, "/cj-icons/gladiator.png", "GLA"),
        Pugilist(2, "/cj-icons/pugilist.png", "PGL"),
        Marauder(3, "/cj-icons/marauder.png", "MRD"),
        Lancer(4, "/cj-icons/lancer.png", "LNC"),
        Archer(5, "/cj-icons/archer.png", "ARC"),
        Conjurer(6, "/cj-icons/conjurer.png", "CNJ"),
        Thaumaturge(7, "/cj-icons/thaumaturge.png", "THM"),
        Arcanist(26, "/cj-icons/arcanist.png", "ACN"),
        Rogue(29, "/cj-icons/rogue.png", "ROG")
    }

    enum class CombatJob(override val id: Int, override val icon: String, override val abbrev: String, val clazz: CombatClass?) : ClassJob {
        Paladin(19, "/cj-icons/paladin.png", "PLD", CombatClass.Gladiator),
        Monk(20, "/cj-icons/monk.png", "MNK", CombatClass.Pugilist),
        Warrior(21, "/cj-icons/warrior.png", "WAR", CombatClass.Marauder),
        Dragoon(22, "/cj-icons/dragoon.png", "DRG", CombatClass.Lancer),
        Bard(23, "/cj-icons/bard.png", "BRD", CombatClass.Archer),
        WhiteMage(24, "/cj-icons/whitemage.png", "WHM", CombatClass.Conjurer),
        BlackMage(25, "/cj-icons/blackmage.png", "BLM", CombatClass.Thaumaturge),
        Summoner(27, "/cj-icons/summoner.png", "SMN", CombatClass.Arcanist),
        Scholar(28, "/cj-icons/scholar.png", "SCH", CombatClass.Arcanist),
        Ninja(30, "/cj-icons/ninja.png", "NIN", CombatClass.Rogue),
        Machinist(31, "/cj-icons/machinist.png", "MCH", null),
        DarkKnight(32, "/cj-icons/darkknight.png", "DRK", null),
        Astrologian(33, "/cj-icons/astrologian.png", "AST", null),
        Samurai(34, "/cj-icons/samurai.png", "SAM", null),
        RedMage(35, "/cj-icons/redmage.png", "RDM", null),
        Gunbreaker(37, "/cj-icons/gunbreaker.png", "GNB", null),
        Dancer(38, "/cj-icons/dancer.png", "DNC", null)
    }

    enum class CombatLimitedJob(override val id: Int, override val icon: String, override val abbrev: String) : ClassJob {
        BlueMage(36, "/cj-icons/bluemage.png", "BLU")
    }

    enum class HandClass(override val id: Int, override val icon: String, override val abbrev: String) : ClassJob {
        Carpenter(8, "/cj-icons/carpenter.png", "CRP"),
        Blacksmith(9, "/cj-icons/blacksmith.png", "BSM"),
        Armorer(10, "/cj-icons/armorer.png", "ARM"),
        Goldsmith(11, "/cj-icons/goldsmith.png", "GSM"),
        Leatherworker(12, "/cj-icons/leatherworker.png", "LTW"),
        Weaver(13, "/cj-icons/weaver.png", "WVR"),
        Alchemist(14, "/cj-icons/alchemist.png", "ALC"),
        Culinarian(15, "/cj-icons/culinarian.png", "CUL"),
    }

    enum class LandClass(override val id: Int, override val icon: String, override val abbrev: String) : ClassJob {
        Miner(16, "/cj-icons/miner.png", "MIN"),
        Botanist(17, "/cj-icons/botanist.png", "BTN"),
        Fisher(18, "/cj-icons/fisher.png", "FSH")
    }
}

File: XivApi/Character/ClassJobs.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Collections.ObjectModel;
using System.Drawing;
using System.Linq;

namespace XivApi.Character
{
    public interface IClassJob : IEquatable<IClassJob>
    {
        int Id { get; }
        Bitmap Icon { get; }
        string Abbrev { get; }

        string Name { get; }
    }

    public abstract class BaseClassJob : IClassJob
    {
        private static readonly Dictionary<int, IClassJob> _classJobs = new Dictionary<int, IClassJob>();

        public static IReadOnlyDictionary<int, IClassJob> ClassJobs { get; } =
            new ReadOnlyDictionary<int, IClassJob>(_classJobs);

        protected BaseClassJob(int id, string abbrev, string name) {
            Id = id;
            Icon = (Bitmap) Resources.ResourceManager.GetObject(name.ToLowerInvariant(), Resources.Culture)!;
            Abbrev = abbrev.ToUpperInvariant();
            Name = name;
            _classJobs[Id] = this;
        }

        public bool Equals(IClassJob? other) {
            if (other == null) throw new ArgumentNullException(nameof(other));
            return Id == other.Id;
        }

        public int Id { get; }
        public Bitmap Icon { get; }
        public string Abbrev { get; }
        public string Name { get; }
    }

    public class CombatClass : BaseClassJob
    {
        internal static void Init() { }
        private CombatClass(int id, string abbrev, string name) : base(id, abbrev, name) { }

        public static readonly CombatClass Gladiator,
            Pugilist,
            Marauder,
            Lancer,
            Archer,
            Conjurer,
            Thaumaturge,
            Arcanist,
            Rogue;

        static CombatClass() {
            Gladiator = new CombatClass(1, "GLA", "Gladiator");
            Pugilist = new CombatClass(2, "PGL", "Pugilist");
            Marauder = new CombatClass(3, "MRD", "Marauder");
            Lancer = new CombatClass(4, "LNC", "Lancer");
            Archer = new CombatClass(5, "ARC", "Archer");
            Conjurer = new CombatClass(6, "CNJ", "Conjurer");
            Thaumaturge = new CombatClass(7, "THM", "Thaumaturge");
            Arcanist = new CombatClass(26, "ACN", "Arcanist");
            Rogue = new CombatClass(29, "ROG", "Rogue");
        }
    }

    public class CombatJob : BaseClassJob
    {
        internal static void Init() { }
        public CombatClass? Class { get; }

        private CombatJob(int id, string abbrev, string name, CombatClass? @class) : base(id, abbrev, name) {
            Class = @class;
        }

        public static readonly CombatJob
            Paladin,
            Monk,
            Warrior,
            Dragoon,
            Bard,
            WhiteMage,
            BlackMage,
            Summoner,
            Scholar,
            Ninja,
            Machinist,
            DarkKnight,
            Astrologian,
            Samurai,
            RedMage,
            Gunbreaker,
            Dancer;

        static CombatJob() {
            Dancer = new CombatJob(38, "DNC", "Dancer", null);
            Gunbreaker = new CombatJob(37, "GNB", "Gunbreaker", null);
            RedMage = new CombatJob(35, "RDM", "RedMage", null);
            Samurai = new CombatJob(34, "SAM", "Samurai", null);
            Astrologian = new CombatJob(33, "AST", "Astrologian", null);
            DarkKnight = new CombatJob(32, "DRK", "DarkKnight", null);
            Machinist = new CombatJob(31, "MCH", "Machinist", null);
            Ninja = new CombatJob(30, "NIN", "Ninja", CombatClass.Rogue);
            Scholar = new CombatJob(28, "SCH", "Scholar", CombatClass.Arcanist);
            Summoner = new CombatJob(27, "SMN", "Summoner", CombatClass.Arcanist);
            BlackMage = new CombatJob(25, "BLM", "BlackMage", CombatClass.Thaumaturge);
            WhiteMage = new CombatJob(24, "WHM", "WhiteMage", CombatClass.Conjurer);
            Bard = new CombatJob(23, "BRD", "Bard", CombatClass.Archer);
            Dragoon = new CombatJob(22, "DRG", "Dragoon", CombatClass.Lancer);
            Warrior = new CombatJob(21, "WAR", "Warrior", CombatClass.Marauder);
            Monk = new CombatJob(20, "MNK", "Monk", CombatClass.Pugilist);
            Paladin = new CombatJob(19, "PLD", "Paladin", CombatClass.Gladiator);
        }
    }

    public class LimitedCombatJob : BaseClassJob
    {
        internal static void Init() { }
        private LimitedCombatJob(int id, string abbrev, string name) : base(id, abbrev, name) { }

        public static readonly LimitedCombatJob
            BlueMage;

        static LimitedCombatJob() {
            BlueMage = new LimitedCombatJob(36, "BLU", "BlueMage");
        }
    }

    public class HandClass : BaseClassJob
    {
        internal static void Init() { }
        private HandClass(int id, string abbrev, string name) : base(id, abbrev, name) { }

        public static readonly HandClass
            Carpenter,
            Blacksmith,
            Armorer,
            Goldsmith,
            Leatherworker,
            Weaver,
            Alchemist,
            Culinarian;

        static HandClass() {
            Culinarian = new HandClass(15, "CUL", "Culinarian");
            Alchemist = new HandClass(14, "ALC", "Alchemist");
            Weaver = new HandClass(13, "WVR", "Weaver");
            Leatherworker = new HandClass(12, "LTW", "Leatherworker");
            Goldsmith = new HandClass(11, "GSM", "Goldsmith");
            Armorer = new HandClass(10, "ARM", "Armorer");
            Blacksmith = new HandClass(9, "BSM", "Blacksmith");
            Carpenter = new HandClass(8, "CRP", "Carpenter");
        }
    }

    public class LandClass : BaseClassJob
    {
        internal static void Init() { }

        private LandClass(int id, string abbrev, string name) : base(id, abbrev, name) { }

        public static readonly LandClass
            Miner,
            Botanist,
            Fisher;

        static LandClass() {
            Fisher = new LandClass(18, "FSH", "Fisher");
            Botanist = new LandClass(17, "BTN", "Botanist");
            Miner = new LandClass(16, "MIN", "Miner");
        }
    }

    public static class ClassJobs
    {
        static ClassJobs() {
            CombatClass.Init();
            CombatJob.Init();
            LimitedCombatJob.Init();
            HandClass.Init();
            LandClass.Init();
            Console.WriteLine("ClassJobs Initialized");
        }

        public static IReadOnlyDictionary<int, IClassJob> ById => BaseClassJob.ClassJobs;

        public static IReadOnlyDictionary<string, IClassJob> ByAbbrev =>
            ById.Values.ToImmutableDictionary(job => job.Abbrev);

        public static IClassJob FindByName(string name) =>
            ById.Values.Single(job => job.Name.ToLowerInvariant().Equals(name.ToLowerInvariant()));
    }
}